src/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs

```diff
@@ -0,0 +1,194 @@
+            while (destination.Length > 0 && (lastResult == OperationStatus.DestinationTooSmall || lastResult == OperationStatus.NeedMoreData))
+            {
+                int bytesConsumed = 0;
+                int bytesWritten = 0;
+
+                if (lastResult == OperationStatus.NeedMoreData)
+                {
+                    int readBytes = 0;
+                    int iter = 0;
+                    while (readBytes < _buffer.Length && ((iter = _stream.Read(_buffer, readBytes, _buffer.Length - readBytes)) > 0))
+                    {
+                        readBytes += iter;
+                        if (readBytes > _buffer.Length)
+                        {
+                            // The stream is either malicious or poorly implemented and returned a number of
+                            // bytes larger than the buffer supplied to it.
+                            throw new InvalidDataException(SR.BrotliStream_Decompress_InvalidStream);
+                        }
+                    }
+                    if (readBytes <= 0)
+                    {
+                        break;
+                    }
+                    source = new Span<byte>(_buffer, 0, readBytes);
+                }
+
+                lastResult = _decoder.Decompress(source, destination, out bytesConsumed, out bytesWritten);
+                if (lastResult == OperationStatus.InvalidData)
+                    throw new InvalidOperationException(SR.BrotliStream_Decompress_InvalidData);
+                if (bytesConsumed > 0)
+                    source = source.Slice(bytesConsumed);
+                if (bytesWritten > 0)
+                {
+                    totalWritten += bytesWritten;
+                    destination = destination.Slice(bytesWritten);
+                }
+            }
+
+            return totalWritten;
+        }
+
+        public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback asyncCallback, object asyncState) =>
+            TaskToApm.Begin(ReadAsync(buffer, offset, count, CancellationToken.None), asyncCallback, asyncState);
+
+        public override int EndRead(IAsyncResult asyncResult) =>
+            TaskToApm.End<int>(asyncResult);
+
+        public override Task<int> ReadAsync(byte[] array, int offset, int count, CancellationToken cancellationToken)
+        {
+            ValidateParameters(array, offset, count);
+            return ReadAsyncMemory(new Memory<byte>(array, offset, count), cancellationToken).AsTask();
+        }
+
+        public override ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (GetType() != typeof(BrotliStream))
+            {
+                // Ensure that existing streams derived from BrotliStream and that override ReadAsync(byte[],...)
+                // get their existing behaviors when the newer Memory-based overload is used.
```

Comment:
Same type check comments as earlier.  I'll stop commenting on it, but it applies to all other such cases.